drivers/dai/intel/alh/alh.c

```diff
@@ -0,0 +1,223 @@
+
+	if (config->channels || config->rate) {
+		alh->params.channels = config->channels;
+		alh->params.rate = config->rate;
+		LOG_INF("%s channels %d rate %d", __func__, config->channels, config->rate);
+	}
+
+	alh->params.stream_id = config->stream_id;
+
+	return 0;
+}
+
+static int dai_alh_set_config_blob(struct dai_intel_alh *dp, const void *spec_config)
+{
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+	const struct dai_intel_ipc4_alh_configuration_blob *blob = spec_config;
+	const struct ipc4_alh_multi_gtw_cfg *alh_cfg = &blob->alh_cfg;
+
+	alh->params.channels = ALH_CHANNELS_DEFAULT;
+	alh->params.rate = ALH_RATE_DEFAULT;
+	/* the LSB 8bits are for stream id */
+	alh->params.stream_id = alh_cfg->mapping[0].alh_id & 0xff;
+
+	return 0;
+}
+
+static int dai_alh_trigger(const struct device *dev, enum dai_dir dir,
+			   enum dai_trigger_cmd cmd)
+{
+	LOG_INF("%s cmd %d", __func__, cmd);
+
+	return 0;
+}
+
+static void alh_claim_ownership(void)
+{
+#if !CONFIG_INTEL_ADSP_CAVS
+	uint32_t ALHASCTL = DT_INST_PROP_BY_IDX(0, reg, 0);
+	uint32_t ALHCSCTL = DT_INST_PROP_BY_IDX(0, reg, 1);
+
+	sys_write32(sys_read32(ALHASCTL) | ALHASCTL_OSEL(0x3), ALHASCTL);
+	sys_write32(sys_read32(ALHCSCTL) | ALHASCTL_OSEL(0x3), ALHCSCTL);
+#endif
+}
+
+static void alh_release_ownership(void)
+{
+#if !CONFIG_INTEL_ADSP_CAVS
+	uint32_t ALHASCTL = DT_INST_PROP_BY_IDX(0, reg, 0);
+	uint32_t ALHCSCTL = DT_INST_PROP_BY_IDX(0, reg, 1);
+
+	sys_write32(sys_read32(ALHASCTL) | ALHASCTL_OSEL(0), ALHASCTL);
+	sys_write32(sys_read32(ALHCSCTL) | ALHASCTL_OSEL(0), ALHCSCTL);
+#endif
+}
+
+
+static const struct dai_config *dai_alh_config_get(const struct device *dev, enum dai_dir dir)
+{
+	struct dai_config *params = (struct dai_config *)dev->config;
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+
+	params->rate = alh->params.rate;
+	params->channels = alh->params.channels;
+	params->word_size = ALH_WORD_SIZE_DEFAULT;
+
+	return params;
+}
+
+static int dai_alh_config_set(const struct device *dev, const struct dai_config *cfg,
+				  const void *bespoke_cfg)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+
+	LOG_INF("%s", __func__);
+
+	alh_claim_ownership();
+
+	if (cfg->type == DAI_INTEL_ALH) {
+		return dai_alh_set_config_tplg(dp, cfg, bespoke_cfg);
+	} else {
+		return dai_alh_set_config_blob(dp, bespoke_cfg);
+	}
+}
+
+static const struct dai_properties *dai_alh_get_properties(const struct device *dev,
+							   enum dai_dir dir, int stream_id)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+	struct dai_properties *prop = &alh->props;
+	uint32_t offset = dir == DAI_DIR_PLAYBACK ?
+		ALH_TXDA_OFFSET : ALH_RXDA_OFFSET;
+
+	prop->fifo_address = dai_base(dp) + offset + ALH_STREAM_OFFSET * stream_id;
+	prop->dma_hs_id = alh_handshake_map[stream_id];
+	prop->stream_id = alh->params.stream_id;
```

Comment:
This will be changed at dai api level, for now it is just using the same way as ssp driver.
For the DMA api it works as you described with user supply the struct to fill and the same we need to apply to dai